### PR TITLE
Java Async APIs

### PR DESCRIPTION
**Pull request by mikaelm12**
Closed https://github.com/aspnet/SignalR/pull/2944 in favor of this.

### CODE REVIEW

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java [hunk 1]
```diff
@@ -150,51 +178,41 @@ public HubConnectionState getConnectionState() {
...
-        }
 
-        logger.log(LogLevel.Debug, "Starting HubConnection");
-        if (transport == null) {
-            transport = new WebSocketTransport(url, logger, headers);
+            transport.setOnReceive(this.callback);
+            return transport.start().thenCompose((future) -> {
+                String handshake = HandshakeProtocol.createHandshakeRequestMessage(new HandshakeRequestMessage(protocol.getName(), protocol.getVersion()));
+                return transport.send(handshake).thenRun(() -> {
+                    hubConnectionState = HubConnectionState.CONNECTED;
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: The connectionState is changed here outside of a lock.

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java [hunk 2]
```diff
@@ -47,21 +48,24 @@ private URI formatUrl(String url) throws URISyntaxException {
     }
 
     @Override
-    public void start() throws Exception {
-        logger.log(LogLevel.Debug, "Starting Websocket connection.");
-        webSocketClient = createWebSocket(headers);
-
-        if (!webSocketClient.connectBlocking()) {
-            String errorMessage = "There was an error starting the Websockets transport.";
-            logger.log(LogLevel.Debug, errorMessage);
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java]: nit: keep this log

**Comment by mikaelm12** [comment]
Updated

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java [hunk 3]
```diff
@@ -211,10 +230,16 @@ private void stop(String errorMessage) {
             logger.log(LogLevel.Debug, "Stopping HubConnection.");
         }
 
-        transport.stop();
-        hubConnectionState = HubConnectionState.DISCONNECTED;
-        connectionState = null;
-        logger.log(LogLevel.Information, "HubConnection stopped.");
+        hubConnectionStateLock.lock();
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: move the lock to the top of `stop`

**Pull request merged**